packages/mui-joy/src/Select/Select.tsx

```diff
@@ -457,19 +471,34 @@ const Select = React.forwardRef(function Select<TValue extends {}>(
   };
 
   const classes = useUtilityClasses(ownerState);
-  const externalForwardedProps = { ...other, slots, slotProps };
+  const externalForwardedProps = {
+    ...other,
+    slots,
+    slotProps,
+  } as SelectSlotsAndSlotProps<boolean> & typeof other;
+
+  const selectedOption = React.useMemo(() => {
+    let selectedOptionsMetadata: SelectValue<SelectOption<OptionValue>, Multiple>;
+    if (multiple) {
+      selectedOptionsMetadata = (value as OptionValue[])
+        .map((v) => getOptionMetadata(v))
+        .filter((o) => o !== undefined) as SelectValue<SelectOption<OptionValue>, Multiple>;
+    } else {
+      selectedOptionsMetadata = (getOptionMetadata(value as OptionValue) ?? null) as SelectValue<
+        SelectOption<OptionValue>,
+        Multiple
+      >;
+    }
 
-  const selectedOption = React.useMemo(
-    () => getOptionMetadata(value as TValue) ?? null,
-    [getOptionMetadata, value],
-  );
+    return selectedOptionsMetadata;
+  }, [getOptionMetadata, value, multiple]);
 
   const [SlotRoot, rootProps] = useSlot('root', {
     ref: handleRef,
     className: classes.root,
     elementType: SelectRoot,
     externalForwardedProps,
-    ownerState,
+    ownerState: ownerState as SelectOwnerState<any, boolean>,
```

Comment:
TypeScript is one of the reasons I try to avoid components with union value type 🥲. It's really hard to get it right.

Reply:
No need to do anything here.